common/src/types.ts

```diff
@@ -0,0 +1,49 @@
+import * as t from 'io-ts'
+
+export interface RequestSchema {
+  readonly parameters?: t.Any
+  readonly query?: t.Any
+  readonly body?: t.Any
+}
+export interface ResponseSchema {
+  readonly body?: t.Any
+}
+export interface MethodSchema {
+  readonly request?: RequestSchema
+  readonly response?: ResponseSchema
+}
+
+export type MethodName = 'all' | 'get' | 'post' | 'put' | 'delete' | 'patch' | 'options' | 'head'
+export type RootPath = '/'
+export type ChildPath = `/${string}`
+
+type Methods = Readonly<Partial<Record<MethodName, MethodSchema>>>
+type Children = Readonly<Partial<Record<Exclude<ChildPath, RootPath>, EndpointSchema>>>
+export interface EndpointSchema extends Methods {
```

Comment:
I think API makes sense for the root-level object, but what about something like **/clubs**? It makes more sense (to me, at least) to call that an endpoint

Maybe I can create a typealias `APISchema = EndpointSchema` so we get the best of both names 🤔

Reply:
> Wait..are `EndpointSchema` and `Children`'s definition mutually recursive?

Yep! The `/clubs` endpoint is also a fully-fledged Endpoint schema. `/clubs/:id` would be yet another layer in, but it also has an Endpoint schema. Altogether, it forms a tree, with our base API as the root.

That's why it's challenging to come up with a good name for things. I wouldn't really call the root API an "endpoint", but it quacks and walks like one. `/clubs/:id` sounds much more like an endpoint to me.

Since I never formally learned web dev, I don't really know what the proper terminology is. Are `/` and `/clubs` and `/club/:id` "endpoints" or "APIs"?

Reply:
> Since I never formally learned web dev, I don't really know what the proper terminology is. Are / and /clubs and /club/:id "endpoints" or "APIs"?

My definition is the "API" is the structure, grouping, and schema of all "endpoints".

Reply:
Ok, then we can stick to "Endpoint".

Following your definition, maybe this makes sense:

```ts
type API = {
  root: Endpoint
}
```

Reply:
Yeah I like that!